src/session-user/session-user.service.ts

```diff
@@ -1,4 +1,119 @@
-import { Injectable } from '@nestjs/common';
+import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import * as _ from 'lodash';
+import { SessionUserEntity } from 'src/entities/session-user.entity';
+import { SessionEntity } from 'src/entities/session.entity';
+import { IFirebaseUser } from 'src/firebase/firebase-user.interface';
+import { SessionService } from 'src/session/session.service';
+import { GetUserDto } from 'src/user/dtos/get-user.dto';
+import { UserRepository } from 'src/user/user.repository';
+import { UserService } from 'src/user/user.service';
+import { STORYBLOK_STORY_STATUS_ENUM } from 'src/utils/constants';
+import { CourseUserService } from '../course-user/course-user.service';
+import { CourseService } from '../course/course.service';
+import { CreateSessionUserDto } from './dtos/create-session-user.dto';
+import { SessionUserRepository } from './session-user.repository';
 
 @Injectable()
-export class SessionUserService {}
+export class SessionUserService {
+  constructor(
+    @InjectRepository(SessionUserRepository) private sessionUserRepository: SessionUserRepository,
+    @InjectRepository(UserRepository) private userRepository: UserRepository,
+    private readonly courseUserService: CourseUserService,
+    private readonly userService: UserService,
+    private readonly sessionService: SessionService,
+    private readonly courseService: CourseService,
+  ) {}
+
+  private markCourseComplete(userCourse: GetUserDto, courseSession: SessionEntity[]): boolean {
+    const userCompletedSession = userCourse.course[0].session;
+
+    const userSessionIds = userCompletedSession.map((session) => {
+      if (session.completed) return session.id;
+    });
+
+    const courseSessionIds = courseSession.map((session) => {
+      if (session.status === STORYBLOK_STORY_STATUS_ENUM.PUBLISHED) return session.id;
+    });
+
+    return _.xor(courseSessionIds, userSessionIds).length == 0;
+  }
+
+  public async createSessionUser(
+    { uid }: IFirebaseUser,
+    { sessionId }: CreateSessionUserDto,
+  ): Promise<SessionUserEntity> {
+    const user = await this.userRepository.findOne({ firebaseUid: uid });
+
+    const { courseId } = await this.sessionService.getSession(sessionId);
+
+    const courseSessions = await this.courseService.getCourseSessions(courseId);
+
+    if (!courseSessions) {
+      throw new HttpException('COURSE SESSIONS NOT FOUND', HttpStatus.NOT_FOUND);
+    }
+
+    let courseUser = await this.courseUserService.getCourseUser({ userId: user.id, courseId });
+
+    if (!courseUser) {
+      courseUser = await this.courseUserService.createCourseUser({ userId: user.id, courseId });
+    }
+
+    const createSessionUserObject = this.sessionUserRepository.create({
```

Comment:
if we remove this const `createSessionUserObject` and just add the dict directly into the `.save` on line 74, i think it'd be a clearer flow and save a couple lines of code

```
   {
      sessionId,
      courseUserId: courseUser.id,
      completed: false,
    }
```